bot/exts/filtering/_filter_lists/filter_list.py

```diff
@@ -0,0 +1,309 @@
+import dataclasses
+import typing
+from abc import abstractmethod
+from collections import defaultdict
+from collections.abc import Iterable
+from dataclasses import dataclass
+from enum import Enum
+from functools import reduce
+from operator import or_
+from typing import Any
+
+import arrow
+from discord.ext.commands import BadArgument
+
+from bot.exts.filtering._filter_context import Event, FilterContext
+from bot.exts.filtering._filters.filter import Filter, UniqueFilter
+from bot.exts.filtering._settings import ActionSettings, Defaults, create_settings
+from bot.exts.filtering._utils import FieldRequiring, past_tense
+from bot.log import get_logger
+
+if typing.TYPE_CHECKING:
+    from bot.exts.filtering.filtering import Filtering
+
+log = get_logger(__name__)
+
+
+class ListType(Enum):
+    """An enumeration of list types."""
+
+    DENY = 0
+    ALLOW = 1
+
+
+#  Alternative names with which each list type can be specified in commands.
+aliases = (
+    (ListType.DENY, {"deny", "blocklist", "blacklist", "denylist", "bl", "dl"}),
+    (ListType.ALLOW, {"allow", "allowlist", "whitelist", "al", "wl"})
+)
+
+
+def list_type_converter(argument: str) -> ListType:
+    """A converter to get the appropriate list type."""
+    argument = argument.lower()
+    for list_type, list_aliases in aliases:
+        if argument in list_aliases or argument in map(past_tense, list_aliases):
+            return list_type
+    raise BadArgument(f"No matching list type found for {argument!r}.")
+
+
+# AtomicList and its subclasses must have eq=False, otherwise the dataclass deco will replace the hash function.
```

Comment:
Why does these have to be hashable in any case? I'm not sure how it makes sense for something to not be comparable (no `__eq__`) but to be hashable.

Reply:
Because I'm using the class' instances as dict keys https://github.com/python-discord/bot/blob/dc0ce05ff1406954f5006366b2899c4bde1967ce/bot/exts/filtering/filtering.py#L912

There's no logic to it being hashable but not comparable beyond not needing to compare.

I don't mind returning `hash(("AtomicList", self.id))` or something of that sort, though I'm not sure how it's preferable.